feature/session/src/main/java/io/github/droidkaigi/confsched2019/session/ui/widget/SessionToolbarBehavior.kt

```diff
@@ -0,0 +1,131 @@
+package io.github.droidkaigi.confsched2019.session.ui.widget
+
+import android.content.Context
+import android.util.AttributeSet
+import android.util.TypedValue
+import android.view.View
+import android.view.animation.AccelerateDecelerateInterpolator
+import android.widget.LinearLayout
+import androidx.appcompat.widget.AppCompatTextView
+import androidx.appcompat.widget.Toolbar
+import androidx.coordinatorlayout.widget.CoordinatorLayout
+import androidx.core.view.ViewCompat
+import androidx.core.view.ViewPropertyAnimatorListener
+import androidx.core.view.children
+import androidx.core.widget.NestedScrollView
+import io.github.droidkaigi.confsched2019.session.R
+
+class SessionToolbarBehavior(
+    val context: Context? = null,
+    attrs: AttributeSet? = null
+) : CoordinatorLayout.Behavior<LinearLayout>(context, attrs) {
+
+    private var sessionTitle: String = ""
+    private var hasSetToolbarTitle = false
+    private var appCompatTextView: AppCompatTextView? = null
+
+    private var mIsAnimation = false
+
+    constructor(context: Context?, attrs: AttributeSet?, sessionTitle: String) : this(
+        context,
+        attrs
+    ) {
+        this.sessionTitle = sessionTitle
+    }
+
+    override fun onStartNestedScroll(
+        coordinatorLayout: CoordinatorLayout,
+        child: LinearLayout,
+        directTargetChild: View,
+        target: View,
+        axes: Int,
+        type: Int
+    ): Boolean = axes == ViewCompat.SCROLL_AXIS_VERTICAL
+
+    override fun onNestedPreScroll(
+        coordinatorLayout: CoordinatorLayout,
+        child: LinearLayout,
+        target: View,
+        dx: Int,
+        dy: Int,
+        consumed: IntArray,
+        type: Int
+    ) {
+        val toolbar = child.findViewById<Toolbar>(R.id.session_toolbar)
+        val nestedScrollView = if (child.childCount > 0) {
+            (child.children.find { it is NestedScrollView } as? NestedScrollView)
+        } else {
+            null
+        }
+        performAnimation(toolbar, nestedScrollView)
+    }
+
+    private fun performAnimation(toolbar: Toolbar, nestedScrollView: NestedScrollView?) {
+        if (!mIsAnimation) {
+            if (nestedScrollView?.canScrollVertically(NEGATIVE_DIRECTION) == false) {
+                appCompatTextView?.let {
+                    animateTitle(TOP_TITLE_ANIMATION_ALPHA, TOP_TITLE_ANIMATION_DURATION, it)
+                    animateToolbarElevation(toolbar, TOP_TOOLBAR_ELEVATION)
+                }
+            } else {
+                initToolbarTitle(toolbar)
+                appCompatTextView?.let {
+                    animateTitle(NO_TOP_TITLE_ANIMATION_ALPHA, NO_TOP_TITLE_ANIMATION_DURATION, it)
+                    animateToolbarElevation(toolbar, NO_TOP_TOOLBAR_ELEVATION)
+                }
+            }
+        }
+    }
+
+    private fun initToolbarTitle(toolbar: Toolbar) {
```

Comment:
This name sounds misleading. 
This method will do:

- Set the title *only once*
- Find a TextView from the toolbar surely *only once*

I'm just curious so let me ask a question. 
https://github.com/DroidKaigi/conference-app-2019/pull/525/files#diff-a45d9aed66a757a483bc9cdf2b196c9aR218 sets the title to this Toolbar (is my understanding correct?) so why is this needed?
I couldn't understand why this class found `toolbar` every `onNestedPreScroll` but this logic would be run only once. Could you please describe the background?

Reply:
> This name sounds misleading.
This method will do:

How about to rename to `setToolbarTitle`
or
rewrite performAction() like below
```
    private fun performAnimation(nestedScrollView: NestedScrollView?) {
        if (!mIsAnimation) {
            if (nestedScrollView?.canScrollVertically(NEGATIVE_DIRECTION) == false) {
                textView?.let {
                    animateTitle(TOP_TITLE_ANIMATION_ALPHA, it)
                    toolbar.elevation = context.resources.getDimension(
                        R.dimen.session_detail_toolbar_elevation_top
                    ) / context.resources.displayMetrics.density
                }
                return
            }
            if (!hasSetToolbarTitle) {
                toolbar.title = sessionTitle
                textView = toolbar.children.find { it is TextView } as TextView
                hasSetToolbarTitle = true
            }
            textView?.let {
                animateTitle(NO_TOP_TITLE_ANIMATION_ALPHA, it)
                toolbar.elevation = context.resources.getDimension(
                    R.dimen.session_detail_toolbar_elevation_not_top
                ) / context.resources.displayMetrics.density
            }
        }
    }
```

> I couldn't understand why this class found toolbar every onNestedPreScroll but this logic would be run only once. Could you please describe the background?

I see. I think that The logic to find `toolbar` isn't needed to call every `onNestedPreScroll`. Therefore, how about that toolbar is passed in constructor of this class.
```
class SessionToolbarBehavior(
    private val context: Context,
    private val toolbar: Toolbar,
    private val sessionTitle: String
) : CoordinatorLayout.Behavior<LinearLayout>() {
```